Fixes #634 Selecting language for voice input

keianhzo opened the pull request:
Fixes #634 #767 Added a new setting to support the new batch of languages. Also rearranged the settings panel.

cvan commented on app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java:
```diff
@@ -305,4 +306,15 @@ public void setMSAALevel(int level) {
 
+    public String getVoiceSearchLanguage() {
+        return mPrefs.getString(
+                mContext.getString(R.string.settings_key_voice_search_language), LocaleUtils.getCurrentLocale());
```
I don't see `voice_search_language` in `strings.xml` in here or in our [`strings.xml`](https://github.com/MozillaReality/FirefoxReality/blob/096424app/src/main/res/values/strings.xml
) in our PR #711 (but I do see it in the [Wiki page](https://github.com/MozillaReality/FirefoxReality/wiki/L10n#voice_search_language) - thanks!)

keianhzo replied:
As per our conversation in Slack we will need to keep those language related strings here for the moment.

Regarding **settings_key_voice_search_language** that's a settings key so not being a translatable string it lives in **non_L10n.xml**.

keianhzo replied:
@cvan can you also provide a longer logcat, I don't seem to be able to reproduce that crash in local. I think the real crash is a bit above the one that you sent.

cvan replied:
happened again on another Oculus Go headset:

https://gist.github.com/cvan/675cff1162dc9f5a4cf8153ab3627be8

not sure what to sift through from that

keianhzo replied:
@cvan The crash should be fixed.

keianhzo commented:
@cvan Updated the buttons order and paddings.

keianhzo commented:
@cvan The crash should be fixed.

keianhzo commented:
@daoshengmu could you do a Chinese review?

cvan commented on app/src/main/res/layout/settings.xml:
```diff
@@ -68,71 +73,79 @@
+                        <org.mozilla.vrbrowser.ui.views.HoneycombButton
+                            android:id="@+id/displayButton"
+                            style="?attr/settingsButtonStyle"
+                            android:layout_marginRight="10dp"
```
can you adjust the vertical centreing of this one?

![image](https://user-images.githubusercontent.com/203725/48245807-9a229e80-e3a1-11e8-9a0b-5f0b78a40eba.png)

it'd be nice if all the icons are lined up on the bottom row of honeycombs too. thanks in advance!

keianhzo replied:
Fixed buttons alignments.

daoshengmu commented:
@keianhzo it seems like something wrong, it can't recognize what I speak for Traditional and Simplified Chinese. Moreover, I think we don't need to put `Zhuyin` and `Pinyin` in the option because they are typing inputs not languages. We should consider only show `Traditional Chinese`, `Simplified Chinese`. Render native text is also a good suggestion, it would be `繁體中文` `简体中文`.

keianhzo commented:
@cvan I have fixed most of the requested changes:
- Button alignments
- Languages in their native texts
- I have also addressed the issue #767

@daoshengmu I have fixed the Zhuyin/Pinyin issue. Can you try again see if works. I've tried with probably the only thing I know and seemed to work but of course that's not a great test.

@andrenatal Can you take a look at what @daoshengmu mentions about chinese recognition? I checked that the language for traditional and simplified chinese are the correct ones but just in case.

daoshengmu commented:
It will crash then I click the mic icon or trying to change the language. I don't have time to trace it. I will take a look at why it crashes on Monday.

keianhzo commented:
@daoshengmu Fixed

andrenatal commented:
Hey guys, sorry I'm on PTO until next Monday and don't have my go with me. I'll check this patch when I get back right after thanksgiving. Thanks!

cvan commented:
@daoshengmu thanks for testing this. that URL issue is significant and a fix (PR #765) has merged into master.

@andrenatal: no worries! we are good on this. will report via email for when you are back. thanks for all your help. Happy Thanksgiving! 🦃

cvan commented:
@keianhzo:

I changed my language to German and did some testing. nice work!

a few questions before merging:

- can you share some step-by-step tips of how to test this?
- which locales have you tested with / which do you need help with?
- how can I tell that the voice commands are correct or incorrect?

keianhzo commented:
@cvan Fixed the merge conflicts, feel free to review and merge and we can address the UI fixes in a follow up so everybody can test this ASAP.

Regarding your questions:

> can you share some step-by-step tips of how to test this?

What I do is:
- Select a new language
- Open voice search and say a recognizable word
- Wait until the search finishes and the word should show up in the URL bar

> which locales have you tested with / which do you need help with?
I've done superficial testing for this languages:
- Spanish
- English
- French
- German
- Chinese
- Italian

Superficial means that I just tested common words, just to make sure that is actually identifying the new language. This will require native speaker testing that's also a good reason for landing it if there is anything outstanding.

> how can I tell that the voice commands are correct or incorrect?
Just check if the world that shows up in the Google search bar is correct.

cvan commented:
My locale is `es-ES` (Spanish). I cannot perform Voice Searches.

* Video: https://vimeo.com/user41954276/review/302177210/ea5eaccb50
* Crash reports: https://gist.github.com/cvan/2c1944a4a3494a4e94101c33491c3a63
* STR:
    1. on fresh install, I click on the Voice Search mic, I accept Oculus permissions, the Voice Search prompt does not show up.
    2. I click on the Voice Search mic icon again.
    3. Stuck at the `!Lo Siento!` screen (`Sorry! I could not understand.`) indefinitely. I waited a good 1 minute+.
    4. Click on the Voice Search mic icon again.
    5. This time, the screen shown is `¿Qué le gustaría buscar en Internet?` (`What would you like to search on the Web?`).
    6. No voice search will work.

keianhzo commented:
@cvan I've fixed the merge conflicts and the prompt fix has already landed. Can you test again see if you can still reproduce the issue. In my tests it's working fine. Anyway this will require a broader testing by as many people as possible and locales as possible.

larsbergstrom commented:
I'd like to echo that if we could get this landed soon, it'd be ideal - we really need to be testing this, as I need to submit a 1.1 build on Monday, so having as much of Friday to test widely with a TC-produced build would be super helpful.
